File: workflows/org.linxira.benchmark-r/src/implementations/fastq_qc_v1.R
# fastq.qc.v1 — independent base-R implementation (exact counter arithmetic).
#
# Field-for-field counterpart of the Rust engine's FASTQ quality-control scan
# (engine/crates/linxira-bio-core/src/fastq.rs); see the Python twin in
# org.linxira.benchmark-python for the prose definitions. Parsing reproduces
# the engine's line-level rules: CRLF tolerance, '@' header with a first
# whitespace-delimited identifier, sequence lines until a '+' separator
# (optionally repeating the identifier), quality lines accumulating to the
# sequence length with bytes in 33..=126, and the engine's error cases
# (no records, truncated record, truncated quality, length mismatch,
# explicit Phred+64 byte below the offset). Counting tracks GC/N
# case-insensitively per cycle (capped at max_cycles) plus file-level
# Phred+33 and Phred+64 Q20/Q30 counters; auto-detection applies Phred+33
# when the minimum quality byte is below 59 and otherwise reports
# "ambiguous" with the engine's warning. No package beyond base R.
#
# The reader and the accumulator are environments (reference semantics) so
# the helpers mutate the same counters, mirroring the engine's single
# QcAccumulator; R's value semantics would silently fork the state.

IMPLEMENTATION <- list(
  capability = "fastq.qc.v1",
  input_roles = c("fastq"),
  # The reader opens through gzfile (magic-byte dispatch); declared gzip
  # inputs are part of the contract.
  input_compression = list(fastq = c("none", "gzip")),
  parameters = c("max_cycles", "quality_encoding"),
  packages = character(0),
  software = function() list(),
  run = function(inputs, parameters) {
    path <- inputs[["fastq"]]
    max_cycles <- fastq_qc_as_count(parameters$max_cycles %||% 500L)
    encoding <- parameters$quality_encoding %||% "auto"
    if (!encoding %in% c("auto", "phred+33", "phred+64")) {
      implementation_error(paste0("unsupported quality encoding: ", encoding))
    }

    lines <- fastq_qc_new_reader(path)
    on.exit(close(lines$connection), add = TRUE)
    state <- fastq_qc_new_state()

    repeat {
      header_line <- fastq_qc_next_line(lines)
      if (is.null(header_line)) break
      record <- state$read_count + 1L
      identifier <- fastq_qc_parse_header(header_line, record, lines$line_number)
      sequence_length <- fastq_qc_parse_sequence(
        lines, identifier, record, state, max_cycles
      )
      fastq_qc_parse_quality(
        lines, record, sequence_length, state, encoding, max_cycles
      )
      state$read_count <- state$read_count + 1L
      state$total_bases <- state$total_bases + sequence_length
      if (is.null(state$min_length) || sequence_length < state$min_length) {
        state$min_length <- sequence_length
      }
      if (sequence_length > state$max_length) state$max_length <- sequence_length
      if (sequence_length > max_cycles) state$cycles_truncated <- TRUE
    }

    if (state$read_count == 0L) implementation_error("FASTQ contains no records")

    if (encoding == "phred+33") {
      quality_encoding <- "phred+33"; offset <- 33L
    } else if (encoding == "phred+64") {
      quality_encoding <- "phred+64"; offset <- 64L
    } else if (state$minimum_quality < 59) {
      quality_encoding <- "phred+33"; offset <- 33L
    } else {
      quality_encoding <- "ambiguous"; offset <- 33L
    }

    warnings_out <- character(0)
    if (quality_encoding == "ambiguous") {
      warnings_out <- c(
        warnings_out,
        paste(
          "quality bytes are compatible with both Phred+33 and legacy Phred+64/Solexa",
          "encodings; metrics use Phred+33 unless quality_encoding is explicitly overridden"
        )
      )
    }
    if (state$cycles_truncated) {
      warnings_out <- c(
        warnings_out,
        sprintf("per-cycle metrics are capped at %d cycle(s)", max_cycles)
      )
    }

    q20_count <- if (offset == 64L) state$q20_64 else state$q20_33
    q30_count <- if (offset == 64L) state$q30_64 else state$q30_33

    list(
      read_count = as.numeric(state$read_count),
      total_bases = as.numeric(state$total_bases),
      min_length = as.numeric(state$min_length),
      max_length = as.numeric(state$max_length),
      mean_length = fastq_qc_ratio(state$total_bases, state$read_count),
      gc_percent = fastq_qc_percent(state$gc_count, state$total_bases),
      n_percent = fastq_qc_percent(state$n_count, state$total_bases),
      mean_quality = fastq_qc_mean_quality(state$quality_sum, state$total_bases, offset),
      q20_percent = fastq_qc_percent(q20_count, state$total_bases),
      q30_percent = fastq_qc_percent(q30_count, state$total_bases),
      quality_encoding = quality_encoding,
      applied_quality_offset = offset,
      per_cycle = unname(Map(
        function(index, cycle) fastq_qc_cycle_metrics(index, cycle, offset),
        seq_along(state$cycles),
        state$cycles
      )),
      warnings = as.list(warnings_out)
    )
  }
)

`%||%` <- function(left, right) if (is.null(left)) right else left

fastq_qc_as_count <- function(value) {
  if (is.null(value) || !is.finite(value) || value < 0) {
    implementation_error("max_cycles must be a non-negative integer")
  }
  count <- suppressWarnings(as.integer(value))
  if (is.na(count)) implementation_error("max_cycles must be a non-negative integer")
  count
}

# gzfile reads both gzip members and plain bytes, like the engine's
# magic-byte dispatch (gzfile(path, "rb") self-opens); readLines splits on
# \n, so one trailing \r is dropped by hand (CRLF tolerance). The reader is
# an environment because the line number must advance across helper calls.
fastq_qc_new_reader <- function(path) {
  reader <- new.env(parent = emptyenv())
  reader$connection <- gzfile(path, "rb")
  reader$line_number <- 0L
  reader
}

fastq_qc_next_line <- function(reader) {
  line <- readLines(reader$connection, n = 1L, warn = FALSE)
  if (length(line) == 0L) return(NULL)
  reader$line_number <- reader$line_number + 1L
  sub("\r$", "", line)
}

fastq_qc_bytes <- function(line) {
  if (!nzchar(line)) return(integer(0))
  as.integer(charToRaw(line))
}

fastq_qc_first_field <- function(bytes) {
  whitespace <- c(32L, 9L, 10L, 13L, 11L, 12L)
  start <- 1L
  while (start <= length(bytes) && bytes[[start]] %in% whitespace) start <- start + 1L
  end <- length(bytes)
  while (end >= start && bytes[[end]] %in% whitespace) end <- end - 1L
  if (end < start) return(NULL)
  slice <- bytes[start:end]
  cut <- which(slice %in% whitespace)
  if (length(cut) > 0L) slice <- slice[seq_len(cut[[1L]] - 1L)]
  if (length(slice) == 0L) return(NULL)
  rawToChar(as.raw(slice))
}

fastq_qc_parse_header <- function(line, record, line_number) {
  bytes <- fastq_qc_bytes(line)
  if (length(bytes) == 0L || bytes[[1L]] != as.integer(charToRaw("@"))) {
    implementation_error(sprintf(
      "malformed FASTQ record %d at line %d: expected a header beginning with '@'",
      record, line_number
    ))
  }
  identifier <- fastq_qc_first_field(bytes[-1L])
  if (is.null(identifier)) {
    implementation_error(sprintf(
      "malformed FASTQ record %d at line %d: header has no identifier",
      record, line_number
    ))
  }
  identifier
}

fastq_qc_new_state <- function() {
  state <- new.env(parent = emptyenv())
  state$read_count <- 0L
  state$total_bases <- 0
  state$min_length <- NULL
  state$max_length <- 0
  state$gc_count <- 0
  state$n_count <- 0
  state$quality_sum <- 0
  state$minimum_quality <- 256L
  state$q20_33 <- 0L
  state$q30_33 <- 0L
  state$q20_64 <- 0L
  state$q30_64 <- 0L
  state$cycles <- list()
  state$cycles_truncated <- FALSE
  state
}

fastq_qc_cycle_slot <- function(state, index) {
  while (length(state$cycles) < index) {
    state$cycles[[length(state$cycles) + 1L]] <- list(
      base_count = 0, gc_count = 0, n_count = 0, quality_sum = 0,
      q20_33 = 0L, q30_33 = 0L, q20_64 = 0L, q30_64 = 0L
    )
  }
  state$cycles[[index]]
}

fastq_qc_add_base <- function(state, cycle_index, byte, max_cycles) {
  upper <- byte
  if (byte == 0x67 || byte == 0x63 || byte == 0x6E) upper <- byte - 0x20  # g c n
  is_gc <- upper == 0x47 || upper == 0x43  # G C
  is_n <- upper == 0x4E                    # N
  if (is_gc) state$gc_count <- state$gc_count + 1
  if (is_n) state$n_count <- state$n_count + 1
  if (cycle_index <= max_cycles) {
    slot <- fastq_qc_cycle_slot(state, cycle_index)
    slot$base_count <- slot$base_count + 1
    if (is_gc) slot$gc_count <- slot$gc_count + 1
    if (is_n) slot$n_count <- slot$n_count + 1
    state$cycles[[cycle_index]] <- slot
  }
  invisible(NULL)
}

fastq_qc_add_quality <- function(state, cycle_index, byte, max_cycles) {
  state$quality_sum <- state$quality_sum + byte
  if (byte < state$minimum_quality) state$minimum_quality <- byte
  if (byte >= 53L) state$q20_33 <- state$q20_33 + 1L
  if (byte >= 63L) state$q30_33 <- state$q30_33 + 1L
  if (byte >= 84L) state$q20_64 <- state$q20_64 + 1L
  if (byte >= 94L) state$q30_64 <- state$q30_64 + 1L
  if (cycle_index <= max_cycles) {
    slot <- fastq_qc_cycle_slot(state, cycle_index)
    slot$quality_sum <- slot$quality_sum + byte
    if (byte >= 53L) slot$q20_33 <- slot$q20_33 + 1L
    if (byte >= 63L) slot$q30_33 <- slot$q30_33 + 1L
    if (byte >= 84L) slot$q20_64 <- slot$q20_64 + 1L
    if (byte >= 94L) slot$q30_64 <- slot$q30_64 + 1L
    state$cycles[[cycle_index]] <- slot
  }
  invisible(NULL)
}

fastq_qc_parse_sequence <- function(lines, identifier, record, state, max_cycles) {
  sequence_length <- 0
  repeat {
    line <- fastq_qc_next_line(lines)
    if (is.null(line)) {
      implementation_error(sprintf(
        "truncated FASTQ record %d at line %d: expected a '+' separator line",
        record, lines$line_number + 1L
      ))
    }
    bytes <- fastq_qc_bytes(line)
    if (length(bytes) > 0L && bytes[[1L]] == as.integer(charToRaw("+"))) {
      if (sequence_length == 0) {
        implementation_error(sprintf(
          "malformed FASTQ record %d at line %d: sequence is empty",
          record, lines$line_number
        ))
      }
      separator_identifier <- fastq_qc_first_field(bytes[-1L])
      if (!is.null(separator_identifier) && !identical(separator_identifier, identifier)) {
        implementation_error(sprintf(
          "malformed FASTQ record %d at line %d: separator identifier does not match the header identifier",
          record, lines$line_number
        ))
      }
      return(sequence_length)
    }
    if (length(bytes) == 0L) {
      implementation_error(sprintf(
        "malformed FASTQ record %d at line %d: sequence line is empty",
        record, lines$line_number
      ))
    }
    for (column in seq_along(bytes)) {
      byte <- bytes[[column]]
      if (byte <= 0x20 || byte >= 0x7F) {
        implementation_error(sprintf(
          "malformed FASTQ record %d at line %d: invalid sequence byte 0x%02x at column %d",
          record, lines$line_number, byte, column
        ))
      }
      fastq_qc_add_base(state, sequence_length + 1L, byte, max_cycles)
      sequence_length <- sequence_length + 1
    }
  }
}

fastq_qc_parse_quality <- function(lines, record, sequence_length, state, encoding, max_cycles) {
  quality_length <- 0
  while (quality_length < sequence_length) {
    line <- fastq_qc_next_line(lines)
    if (is.null(line)) {
      implementation_error(sprintf(
        "truncated FASTQ record %d at line %d: sequence length is %d, but only %d quality values were present",
        record, lines$line_number + 1L, sequence_length, quality_length
      ))
    }
    bytes <- fastq_qc_bytes(line)
    if (length(bytes) == 0L) {
      implementation_error(sprintf(
        "malformed FASTQ record %d at line %d: quality line is empty",
        record, lines$line_number
      ))
    }
    resulting_length <- quality_length + length(bytes)
    if (resulting_length > sequence_length) {
      implementation_error(sprintf(
        "malformed FASTQ record %d at line %d: sequence length is %d, but quality length is %d",
        record, lines$line_number, sequence_length, resulting_length
      ))
    }
    for (column in seq_along(bytes)) {
      byte <- bytes[[column]]
      if (byte < 33L || byte > 126L) {
        implementation_error(sprintf(
          "malformed FASTQ record %d at line %d: invalid quality byte 0x%02x at column %d",
          record, lines$line_number, byte, column
        ))
      }
      if (encoding == "phred+64" && byte < 64L) {
        implementation_error(sprintf(
          "malformed FASTQ record %d at line %d: quality byte 0x%02x at column %d is below the Phred+64 offset",
          record, lines$line_number, byte, column
        ))
      }
      fastq_qc_add_quality(state, quality_length + column, byte, max_cycles)
    }
    quality_length <- resulting_length
  }
  invisible(NULL)
}

fastq_qc_cycle_metrics <- function(index, cycle, quality_offset) {
  q20_count <- if (quality_offset == 64L) cycle$q20_64 else cycle$q20_33
  q30_count <- if (quality_offset == 64L) cycle$q30_64 else cycle$q30_33
  list(
    cycle = index,
    base_count = as.numeric(cycle$base_count),
    gc_percent = fastq_qc_percent(cycle$gc_count, cycle$base_count),
    n_percent = fastq_qc_percent(cycle$n_count, cycle$base_count),
    mean_quality = fastq_qc_mean_quality(cycle$quality_sum, cycle$base_count, quality_offset),
    q20_percent = fastq_qc_percent(q20_count, cycle$base_count),
    q30_percent = fastq_qc_percent(q30_count, cycle$base_count)
  )
}

fastq_qc_ratio <- function(numerator, denominator) {
  if (denominator == 0) 0 else numerator / denominator
}

fastq_qc_percent <- function(numerator, denominator) {
  fastq_qc_ratio(numerator, denominator) * 100
}

fastq_qc_mean_quality <- function(quality_sum, base_count, quality_offset) {
  if (base_count == 0) 0 else quality_sum / base_count - quality_offset
}

File: workflows/org.linxira.benchmark-python/src/implementations/fastq_qc_v1.py
"""fastq.qc.v1 — independent stdlib implementation (exact counter arithmetic).

Counterpart of the Rust engine's FASTQ quality-control scan
(engine/crates/linxira-bio-core/src/fastq.rs). The parser reproduces the
engine's line-level rules exactly:

* CRLF line endings are tolerated (one trailing ``\\r`` is dropped);
* the header must begin with ``@`` and its first whitespace-delimited field
  is the identifier; a ``+`` separator may repeat it (mismatch = error);
* the sequence spans every line until the ``+`` separator line, each byte
  must be ASCII graphic, and an empty sequence is rejected;
* quality lines accumulate until the sequence length is reached, each byte
  must lie in 33..=126, overflow past the sequence length is rejected, and
  an explicit Phred+64 mode rejects bytes below the offset;
* an input with no records is rejected.

Counting follows the engine: GC and N are counted case-insensitively on the
uppercase base, per-cycle metrics are capped at ``max_cycles`` cycles, and
both Phred+33 and Phred+64 Q20/Q30 counters are tracked for every byte.
Encoding auto-detection is the engine's rule: an observed minimum quality
byte below 59 selects Phred+33; otherwise the encoding is reported as
``ambiguous`` with a warning (Phred+33 semantics still apply). All
arithmetic is exact counter arithmetic — no third-party library.
"""

from __future__ import annotations

import gzip
import sys
from pathlib import Path
from typing import Any

CAPABILITY = "fastq.qc.v1"
INPUT_ROLES = ("fastq",)
# The reader dispatches on gzip magic bytes; declared gzip inputs are part
# of the contract (mirrors the rust engine's transparent decompression).
INPUT_COMPRESSION = {"fastq": ("none", "gzip")}
PARAMETERS: tuple[str, ...] = ("max_cycles", "quality_encoding")
GZIP_MAGIC = b"\x1f\x8b"
DEFAULT_MAX_CYCLES = 500


class ImplementationError(ValueError):
    """An input the native engine would also reject, reported the same way."""


def software() -> list[dict[str, str]]:
    return [{"name": "CPython", "version": sys.version.split()[0]}]


def run(inputs: dict[str, Any], parameters: dict[str, Any]) -> dict[str, Any]:
    path = Path(inputs["fastq"])
    max_cycles = parameters.get("max_cycles", DEFAULT_MAX_CYCLES)
    if isinstance(max_cycles, bool) or not isinstance(max_cycles, int) or max_cycles < 0:
        raise ImplementationError("max_cycles must be a non-negative integer")
    encoding = parameters.get("quality_encoding", "auto")
    if encoding not in ("auto", "phred+33", "phred+64"):
        raise ImplementationError(f"unsupported quality encoding: {encoding}")

    with _open_bytes(path) as handle:
        return _scan(handle, max_cycles, encoding)


def _open_bytes(path: Path):
    with path.open("rb") as probe:
        magic = probe.read(2)
    if magic == GZIP_MAGIC:
        return gzip.open(path, "rb")
    return path.open("rb")


class _Cycle:
    """Per-cycle counters (the engine's ``CycleAccumulator``)."""

    __slots__ = ("base_count", "gc_count", "n_count", "quality_sum", "q20_33", "q30_33", "q20_64", "q30_64")

    def __init__(self) -> None:
        self.base_count = 0
        self.gc_count = 0
        self.n_count = 0
        self.quality_sum = 0
        self.q20_33 = 0
        self.q30_33 = 0
        self.q20_64 = 0
        self.q30_64 = 0


class _State:
    """File-level counters (the engine's ``QcAccumulator``)."""

    def __init__(self) -> None:
        self.read_count = 0
        self.total_bases = 0
        self.min_length: int | None = None
        self.max_length = 0
        self.gc_count = 0
        self.n_count = 0
        self.quality_sum = 0
        self.minimum_quality = 256
        self.q20_33 = 0
        self.q30_33 = 0
        self.q20_64 = 0
        self.q30_64 = 0
        self.cycles: list[_Cycle] = []
        self.cycles_truncated = False

    def add_base(self, cycle: int, byte: int, max_cycles: int) -> None:
        upper = byte - 0x20 if byte in (0x67, 0x63, 0x6E) else byte  # g c n → G C N
        if upper in (0x47, 0x43):  # G C
            self.gc_count += 1
        elif upper == 0x4E:  # N
            self.n_count += 1
        if cycle < max_cycles:
            cycle_state = self._cycle(cycle)
            cycle_state.base_count += 1
            if upper in (0x47, 0x43):
                cycle_state.gc_count += 1
            elif upper == 0x4E:
                cycle_state.n_count += 1

    def add_quality(self, cycle: int, byte: int, max_cycles: int) -> None:
        self.quality_sum += byte
        if byte < self.minimum_quality:
            self.minimum_quality = byte
        if byte >= 53:
            self.q20_33 += 1
        if byte >= 63:
            self.q30_33 += 1
        if byte >= 84:
            self.q20_64 += 1
        if byte >= 94:
            self.q30_64 += 1
        if cycle < max_cycles:
            cycle_state = self._cycle(cycle)
            cycle_state.quality_sum += byte
            if byte >= 53:
                cycle_state.q20_33 += 1
            if byte >= 63:
                cycle_state.q30_33 += 1
            if byte >= 84:
                cycle_state.q20_64 += 1
            if byte >= 94:
                cycle_state.q30_64 += 1

    def _cycle(self, index: int) -> _Cycle:
        while len(self.cycles) <= index:
            self.cycles.append(_Cycle())
        return self.cycles[index]


def _scan(handle, max_cycles: int, encoding: str) -> dict[str, Any]:
    state = _State()
    lines = _LineReader(handle)
    while True:
        header_line = lines.next_line()
        if header_line is None:
            break
        record = state.read_count + 1
        identifier = _parse_header(header_line, record, lines.line_number)
        sequence_length = _parse_sequence(lines, identifier, record, state, max_cycles)
        _parse_quality(lines, record, sequence_length, state, encoding, max_cycles)

        state.read_count += 1
        state.total_bases += sequence_length
        if state.min_length is None or sequence_length < state.min_length:
            state.min_length = sequence_length
        state.max_length = max(state.max_length, sequence_length)
        if sequence_length > max_cycles:
            state.cycles_truncated = True

    if state.read_count == 0:
        raise ImplementationError("FASTQ contains no records")

    if encoding == "phred+33":
        quality_encoding, applied_quality_offset = "phred+33", 33
    elif encoding == "phred+64":
        quality_encoding, applied_quality_offset = "phred+64", 64
    elif state.minimum_quality < 59:
        quality_encoding, applied_quality_offset = "phred+33", 33
    else:
        quality_encoding, applied_quality_offset = "ambiguous", 33

    warnings: list[str] = []
    if quality_encoding == "ambiguous":
        warnings.append(
            "quality bytes are compatible with both Phred+33 and legacy Phred+64/Solexa "
            "encodings; metrics use Phred+33 unless quality_encoding is explicitly overridden"
        )
    if state.cycles_truncated:
        warnings.append(f"per-cycle metrics are capped at {max_cycles} cycle(s)")

    use_phred64 = applied_quality_offset == 64
    q20_count = state.q20_64 if use_phred64 else state.q20_33
    q30_count = state.q30_64 if use_phred64 else state.q30_33

    return {
        "read_count": state.read_count,
        "total_bases": state.total_bases,
        "min_length": state.min_length,
        "max_length": state.max_length,
        "mean_length": _ratio(state.total_bases, state.read_count),
        "gc_percent": _percent(state.gc_count, state.total_bases),
        "n_percent": _percent(state.n_count, state.total_bases),
        "mean_quality": _mean_quality(state.quality_sum, state.total_bases, applied_quality_offset),
        "q20_percent": _percent(q20_count, state.total_bases),
        "q30_percent": _percent(q30_count, state.total_bases),
        "quality_encoding": quality_encoding,
        "applied_quality_offset": applied_quality_offset,
        "per_cycle": [
            _cycle_metrics(index, cycle, applied_quality_offset)
            for index, cycle in enumerate(state.cycles)
        ],
        "warnings": warnings,
    }


class _LineReader:
    """Byte lines with one trailing ``\\n`` and one trailing ``\\r`` dropped, like the engine."""

    def __init__(self, handle):
        self._handle = handle
        self.line_number = 0

    def next_line(self) -> bytes | None:
        line = self._handle.readline()
        if line == b"":
            return None
        self.line_number += 1
        if line.endswith(b"\n"):
            line = line[:-1]
        if line.endswith(b"\r"):
            line = line[:-1]
        return line

    def next_expected_line(self) -> int:
        return self.line_number + 1


def _parse_header(line: bytes, record: int, line_number: int) -> bytes:
    if not line.startswith(b"@"):
        raise ImplementationError(
            _malformed(record, line_number, "expected a header beginning with '@'")
        )
    identifier = _first_ascii_field(line[1:])
    if identifier is None:
        raise ImplementationError(_malformed(record, line_number, "header has no identifier"))
    return identifier


def _parse_sequence(
    lines: _LineReader,
    identifier: bytes,
    record: int,
    state: _State,
    max_cycles: int,
) -> int:
    sequence_length = 0
    while True:
        line = lines.next_line()
        if line is None:
            raise ImplementationError(
                f"truncated FASTQ record {record} at line {lines.next_expected_line()}: "
                "expected a '+' separator line"
            )
        if line.startswith(b"+"):
            if sequence_length == 0:
                raise ImplementationError(
                    _malformed(record, lines.line_number, "sequence is empty")
                )
            _validate_separator(line[1:], identifier, record, lines.line_number)
            return sequence_length
        if not line:
            raise ImplementationError(
                _malformed(record, lines.line_number, "sequence line is empty")
            )
        for column, byte in enumerate(line):
            if not (0x20 < byte < 0x7F):
                raise ImplementationError(
                    _malformed(
                        record,
                        lines.line_number,
                        f"invalid sequence byte 0x{byte:02x} at column {column + 1}",
                    )
                )
            state.add_base(sequence_length, byte, max_cycles)
            sequence_length += 1


def _validate_separator(separator: bytes, identifier: bytes, record: int, line: int) -> None:
    separator_identifier = _first_ascii_field(separator)
    if separator_identifier is not None and separator_identifier != identifier:
        raise ImplementationError(
            _malformed(record, line, "separator identifier does not match the header identifier")
        )


def _parse_quality(
    lines: _LineReader,
    record: int,
    sequence_length: int,
    state: _State,
    encoding: str,
    max_cycles: int,
) -> None:
    quality_length = 0
    while quality_length < sequence_length:
        line = lines.next_line()
        if line is None:
            raise ImplementationError(
                f"truncated FASTQ record {record} at line {lines.next_expected_line()}: "
                f"sequence length is {sequence_length}, but only {quality_length} "
                "quality values were present"
            )
        if not line:
            raise ImplementationError(
                _malformed(record, lines.line_number, "quality line is empty")
            )
        resulting_length = quality_length + len(line)
        if resulting_length > sequence_length:
            raise ImplementationError(
                _malformed(
                    record,
                    lines.line_number,
                    f"sequence length is {sequence_length}, but quality length is {resulting_length}",
                )
            )
        for column, byte in enumerate(line):
            if not (33 <= byte <= 126):
                raise ImplementationError(
                    _malformed(
                        record,
                        lines.line_number,
                        f"invalid quality byte 0x{byte:02x} at column {column + 1}",
                    )
                )
            if encoding == "phred+64" and byte < 64:
                raise ImplementationError(
                    _malformed(
                        record,
                        lines.line_number,
                        f"quality byte 0x{byte:02x} at column {column + 1} is below the Phred+64 offset",
                    )
                )
            state.add_quality(quality_length + column, byte, max_cycles)
        quality_length = resulting_length


def _cycle_metrics(index: int, cycle: _Cycle, quality_offset: int) -> dict[str, Any]:
    use_phred64 = quality_offset == 64
    q20_count = cycle.q20_64 if use_phred64 else cycle.q20_33
    q30_count = cycle.q30_64 if use_phred64 else cycle.q30_33
    return {
        "cycle": index + 1,
        "base_count": cycle.base_count,
        "gc_percent": _percent(cycle.gc_count, cycle.base_count),
        "n_percent": _percent(cycle.n_count, cycle.base_count),
        "mean_quality": _mean_quality(cycle.quality_sum, cycle.base_count, quality_offset),
        "q20_percent": _percent(q20_count, cycle.base_count),
        "q30_percent": _percent(q30_count, cycle.base_count),
    }


def _first_ascii_field(line: bytes) -> bytes | None:
    whitespace = b" \t\n\r\x0b\x0c"
    start = 0
    end = len(line)
    while start < end and line[start] in whitespace:
        start += 1
    while end > start and line[end - 1] in whitespace:
        end -= 1
    if start == end:
        return None
    for index in range(start, end):
        if line[index] in whitespace:
            return line[start:index]
    return line[start:end]


def _malformed(record: int, line: int, message: str) -> str:
    return f"malformed FASTQ record {record} at line {line}: {message}"


def _ratio(numerator: int, denominator: int) -> float:
    if denominator == 0:
        return 0.0
    return numerator / denominator


def _percent(numerator: int, denominator: int) -> float:
    return _ratio(numerator, denominator) * 100.0


def _mean_quality(quality_ascii_sum: int, base_count: int, quality_offset: int) -> float:
    if base_count == 0:
        return 0.0
    return quality_ascii_sum / base_count - quality_offset
